rasa/core/policies/ensemble.py

```diff
@@ -1,525 +1,145 @@
-                f"stories to your training data or "
-                f"remove the rule-based policy (`{RulePolicy.__name__}`) from your "
-                f"your pipeline.",
-                docs=DOCS_URL_RULES,
-            )
-        elif not is_rules_consuming_policy_available and contain_rule_trackers:
-            rasa.shared.utils.io.raise_warning(
-                f"Found rule-based training data but no policy supporting rule-based "
-                f"data. Please add `{RulePolicy.__name__}` or another rule-supporting "
-                f"policy to the `policies` section in `{DEFAULT_CONFIG_PATH}`.",
-                docs=DOCS_URL_RULES,
-            )
 
-    def train(
-        self,
-        training_trackers: List[TrackerWithCachedStates],
-        domain: Domain,
-        interpreter: NaturalLanguageInterpreter,
-        **kwargs: Any,
-    ) -> None:
-        if training_trackers:
-            self._emit_rule_policy_warning(training_trackers)
-
-            for policy in self.policies:
-                trackers_to_train = SupportedData.trackers_for_policy(
-                    policy, training_trackers
-                )
-                policy.train(
-                    trackers_to_train, domain, interpreter=interpreter, **kwargs
-                )
+class PolicyPredictionEnsemble:
+    """Interface for any policy prediction ensemble.
 
-            self.action_fingerprints = training.create_action_fingerprints(
-                training_trackers, domain
-            )
-            # set rule only data after training in order to make ensemble usable
-            # without loading
-            self._set_rule_only_data()
-        else:
-            logger.info("Skipped training, because there are no training samples.")
+    Given a list of predictions from policies, which include some meta data about the
+    policies themselves, an "ensemble" decides what the final prediction should be, in
+    the following way:
+    1. If the previously predicted action was rejected, then the ensemble sets the
+       probability for this action to 0.0 (in all given predictions).
+    2. It combines the information from the single predictions, which include some
+       meta data about the policies (e.g. priority), into a final prediction.
+    3. If the sequence of events given at the time of prediction ends with a user
+       utterance, then the ensemble adds a special event to the event-list included in
+       the final prediction that indicates whether the final prediction was made based
+       on the actual text of that user utterance.
 
-        self.date_trained = datetime.now().strftime("%Y%m%d-%H%M%S")
+    Observe that policies predict "mandatory" as well as "optional"
+    events. The ensemble decides which of the optional events should
+    be passed on.
+    """
 
-    def probabilities_using_best_policy(
+    @abstractmethod
+    def combine_predictions(
         self,
+        predictions: List[PolicyPrediction],
```

Comment:
Just realized something: Unless we do a separate `GraphComponent` to merge the predictions into a list or we do some magic in the `GraphComponent` then we can't have several `GraphComponents` return things for the same input params of another one. An easy way to would be to solve this via `kwargs` (and maybe filter for `isinstance(value, PolicyPrediction)`

Reply:
oh, I missed that, yep 😬  thanks!

Reply:
Added a new "combine_predictions_from_kwargs" that just calls the other function. And although the motivation for this function is really the graph, I decided not to make the base class a `GraphComponent`, but maybe we should 🤔 . What do you think?

Reply:
Tbh, the base class currently has no use at all 🤷🏻 (it never had - I think it was ought to be customizable but it de facto isn't 🙈 ). Up to you if you want to merge them as part of this as part of this PR

Reply:
I'll leave the base class here -- I think we could use some separate ensembles (instead of trying to squeeze all "reasonable" behaviour into one ensemble). And judging from the discussions on RulePolicy vs TED vs things extracted from regex interpreter - I think this could happen soon 😅